src/allmydata/test/test_storage.py

```diff
@@ -553,8 +554,14 @@ def call_get_disk_stats(whichdir, reserved_space=0):
 
         # we abandon the first set, so their provisional allocation should be
         # returned
+
+        # XXX okay, so the weak-key-dictionary in storage-server is
+        # basically making an interface to the storage-server that is
+        # "whenever I drop my object AND the garbage-collector removes
+        # it, *then* that thing is no longer writing"..? ... yuuuuck.
```

Comment:
Sounds like it's worth a bug report in the issuer tracker.  That's terrible.

Reply:
Yeah. I didn't look deeply enough to suggest something better, but I know this isn't the right answer :/

As to the wider question, I will file an issue for ^ and reference that instead of the XXX comment about. I'd be tempted to then land this as-is (since arguably it makes pypy behave mostly-like CPython will, at least for the tests) but I guess on the other hand that's "**just** for the tests" and this may be merely hiding other hard-to-find weirdness in the behavior under PyPy...

Reply:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3269#ticket